File: 백준/Silver/1927. 최소 힙/최소 힙.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        
        while(N-- > 0) {
            int x = Integer.parseInt(br.readLine());
            if(x != 0) {
                minHeap.add(x);
                continue;
            }
            if(minHeap.isEmpty()) {
                bw.write(0 + "\n");
            } else {
                bw.write(minHeap.poll() + "\n");
            }
        }
        bw.flush();
        bw.close(); 
        br.close();
    }
}

File: 백준/Silver/1927. 최소 힙/최소 힙.kt
import java.util.*
import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val minHeap = PriorityQueue<Int>()
    var N = br.readLine().toInt()
    while(N-- > 0) {
        val x = br.readLine().toInt()
        if(x != 0) {
            minHeap.add(x)
            continue
        }
        println(if(minHeap.isEmpty()) 0 else minHeap.poll())
    }
}